kong/plugins/log-serializers/runscope.lua

```diff
@@ -0,0 +1,43 @@
+local _M = {}
+
+function _M.serialize(ngx)
+  
+  -- JSON format below based on Runscope Messages resource:
+  -- https://www.runscope.com/docs/api/messages#message-create
+  -- Not sent: req/res:body_encoding, req:form, res:reason
+
+  -- timers
+  -- @see core.handler for their definition
+--  local req_send_time = ngx.ctx.KONG_PROXY_LATENCY or -1
+--  local req_wait_time = ngx.ctx.KONG_WAITING_TIME or -1
+--  local req_receive_time = ngx.ctx.KONG_RECEIVE_TIME or -1
+
+  -- Compute the total time. If some properties were unavailable
+  -- (because the proxying was aborted), then don't add the value.
+----  local req_time = 0
+--  for _, timer in ipairs({req_send_time, req_wait_time, req_receive_time}) do
+--    if timer > 0 then
+  --    req_time = req_time + timer
+  --  end
+--  end
+  return {
+    request = {
+      url = ngx.var.scheme.."://"..ngx.var.host..":"..ngx.var.server_port..ngx.var.request_uri,
+      method = ngx.req.get_method(),
+      headers = ngx.req.get_headers(),
+      body = ngx.ctx.runscope.req_body,
+      timestamp = ngx.req.start_time(),
+      form = ngx.ctx.runscope.req_post_args
+    },
+    response = {
+      status = ngx.status,
+      headers = ngx.resp.get_headers(),
+      size_bytes = ngx.var.bytes_sent,
+      body = ngx.ctx.runscope.res_body,
+      timestamp = ngx.now(),
+      response_time = ngx.now() - ngx.req.start_time()
```

Comment:
### `response_time`
ngx.var.request_time will be the same value (seconds with a `ms` resolution) but already computed. It would just need to be converted to seconds.

One concern though: it will also include the time taken by Kong to send the response to the client. If we want to monitor the time it took for Kong to just receive the (full) response from upstream: `(ngx.ctx.KONG_RECEIVE_TIME or 0) + (ngx.ctx.KONG_WAITING_TIME or 0)` is probably what we want (they are in `ms` so if seconds if required, some conversion must also take place).

### `timestamp`

`ngx.now()` will not return a UNIX timestamp but a floating point number in seconds with a milliseconds decimal part. Some conversion must take place here too.

### `response.size_bytes`

It is said in the Runscope documentation:

> The size of the response body in bytes.

Hence, the correct variable here is `ngx.var.body_bytes_sent`, since `body_bytes` also include the headers size. Also: this will not be the size of the response body *received*, but the one *sent* to the client. Meaning that if any other plugin *modifies* the response body, the size will be inaccurate. This is a limitation to be aware of, thought the size could also be computed manually from `ngx.ctx.runscope.res_body`.

### `request.url`

The same applies here as to `response.body`: if a plugin modified the request's URI, the changes will not be reflected, this will be the request that Kong initially received, not the request made to the upstream API.

---

Overall, I think what matters more here is that the "client" of the API really is Kong, since you want to monitor how your API is doing by itself, and do as if Kong was just a regular client, right? To get metrics on how *Kong itself* performs for your API 9or Kong + API latencies for example), that would be a different tool than Runscope, is the way I see this for now. Hence why I am proposing those changes, so the plugin can most accurately achieve Runscope's goal.